org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/NoGrammarConstraintsCodeAction.java

```diff
@@ -153,18 +134,94 @@ static String getFileName(String schemaURI) {
 		return new File(schemaURI).getName();
 	}
 
-	private static CodeAction createGrammarFileAndBindIt(String title, String grammarURI, String grammarContent,
-			String insertText, int insertOffset, DOMDocument document, Diagnostic diagnostic)
-			throws BadLocationException {
-		Position position = document.positionAt(insertOffset);
-		TextDocumentEdit insertEdit = CodeActionFactory.insertEdit(insertText, position, document.getTextDocument());
-		return createGrammarFileAndBindIt(title, grammarURI, grammarContent, insertEdit, diagnostic);
-	}
-
 	private static CodeAction createGrammarFileAndBindIt(String title, String grammarURI, String grammarContent,
 			TextDocumentEdit boundEdit, Diagnostic diagnostic) {
 		CodeAction codeAction = CodeActionFactory.createFile(title, grammarURI, grammarContent, diagnostic);
 		codeAction.getEdit().getDocumentChanges().add(Either.forLeft(boundEdit));
 		return codeAction;
 	}
+
+	public static TextDocumentEdit createXSINoNamespaceSchemaLocationEdit(String schemaFileName, DOMDocument document)
+			throws BadLocationException {
+		String delimiter = document.getTextDocument().lineDelimiter(0);
+		DOMElement documentElement = document.getDocumentElement();
+		int beforeTagOffset = documentElement.getStartTagOpenOffset();
+		int afterTagOffset = beforeTagOffset + 1 + documentElement.getTagName().length();
+
+		StringBuilder insertText = new StringBuilder();
+		
+		insertText.append(" xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\"");
+		insertText.append(delimiter);
+
+		insertText.append(" xsi:noNamespaceSchemaLocation=\"");
+		insertText.append(schemaFileName);
+		insertText.append("\"");
+
+		Position position = document.positionAt(afterTagOffset);
+		return CodeActionFactory.insertEdit(insertText.toString(), position, document.getTextDocument());
+	}
+
+	public static TextDocumentEdit createXSISchemaLocationEdit(String schemaFileName, String targetNamespace,
+			DOMDocument document) throws BadLocationException {
+		String delimiter = document.getTextDocument().lineDelimiter(0);
+		DOMElement documentElement = document.getDocumentElement();
+		int beforeTagOffset = documentElement.getStartTagOpenOffset();
+		int afterTagOffset = beforeTagOffset + 1 + documentElement.getTagName().length();
+
+		StringBuilder insertText = new StringBuilder();
+		insertText.append(" xmlns=\"");
+		insertText.append(targetNamespace);
+		insertText.append("\"");
+		insertText.append(delimiter);
+		
+		insertText.append(" xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\"");
+		insertText.append(delimiter);
+
+		insertText.append(" xsi:schemaLocation=\"");

```

Comment:
It would be nice to respect `xml.format.xsiSchemaLocationSplit` here.

Reply:
I totally agree with you! I would like to see that in an another PR and try to start new formatter with scanner (instead of DOM).

Reply:
See #1052